packages/core-data/src/hooks/use-entity-record.js

```diff
@@ -0,0 +1,110 @@
+ */
+import { useQuerySelect } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import { store as coreStore } from '../';
+import { IDLE, SUCCESS, ERROR, RESOLVING } from './constants';
+
+/**
+ * @typedef {IDLE|RESOLVING|SUCCESS|ERROR} EntityRecordStatus
+ */
+
+/**
+ * @typedef {Object} EntityRecordResolution
+ * @property {Object}             record       the requested entity record
+ * @property {Object}             editedRecord the requested entity record with any edits applied
+ * @property {boolean}            isMissing    is the record missing after the resolver has finished?
+ * @property {boolean}            isResolving  is the record still being resolved?
+ * @property {boolean}            hasResolved  is the record resolved by now?
+ * @property {boolean}            hasEdits     were there eny edits applied to this entity record?
+ * @property {EntityRecordStatus} status       resolution status
+ */
+
+/**
+ * Resolves the specified entity record.
+ *
+ * @param {string} kind     Kind of the deleted entity.
+ * @param {string} name     Name of the deleted entity.
+ * @param {string} recordId Record ID of the deleted entity.
+ *
+ * @example
+ * ```js
+ * import { useEntityRecord } from '@wordpress/core-data';
+ *
+ * function HammerPriceDisplay( { hammerId } ) {
+ *   const { record, isResolving } = useEntityRecord( 'my-shop', 'hammer', hammerId );
+ *
+ *   if ( isResolving ) {
+ *     return 'Loading...';
+ *   }
+ *
+ *   return new Intl.NumberFormat( 'en-US', {
+ *     style: 'USD',
+ *     currency,
+ *   } ).format( record.price );
+ * }
+ *
+ * // Rendered in the application:
+ * // <HammerPriceDisplay hammerId={ 1 } />
+ * ```
+ *
+ * In the above example, when `HammerPriceDisplay` is rendered into an
+ * application, the price and the resolution details will be retrieved from
+ * the store state using `getEntityRecord()`, or resolved if missing.
+ *
+ * @return {EntityRecordResolution} Entity record data.
+ */
+export default function __experimentalUseEntityRecord( kind, name, recordId ) {
+	const {
+		data,
+		isMissing,
+		isResolving,
+		hasResolved,
+		editedRecord,
+		hasEdits,
+	} = useQuerySelect(
+		( query ) => {
+			const {
+				getEntityRecord,
+				getEditedEntityRecord,
+				hasEditsForEntityRecord,
+			} = query( coreStore );
+			const args = [ kind, name, recordId ];
+			const recordResponse = getEntityRecord( ...args );
+			return {
+				...recordResponse,
+				isMissing: recordResponse.hasResolved && ! recordResponse.data,
+				editedRecord: getEditedEntityRecord( ...args ).data,
+				hasEdits: hasEditsForEntityRecord( ...args ).data,
+			};
+		},
+		[ kind, name, recordId ]
+	);
+
+	let status;
+	if ( isResolving ) {
+		status = RESOLVING;
+	} else if ( hasResolved ) {
+		if ( data ) {
+			status = SUCCESS;
+		} else {
+			status = ERROR;
+		}
+	} else {
+		status = IDLE;
+	}
+
+	return {
+		status,
+		record: data,
+		editedRecord,
+		hasEdits,
+		isMissing,
+		isResolving,
+		hasResolved,
```

Comment:
I think we need to also expose `error` for error handling. Or if it's handled in `data` then I think it makes sense to separate them into different variables.

Reply:
I mean the error details. If we don't have access to the error then I think we should definitely add it 😆 , not in this PR though.

Reply:
Agreed! It is a quite old problem, too – it would be great to get it fixed:

https://github.com/WordPress/gutenberg/blob/70730e61ba3148ebfa2be125d17c8138ab16392c/packages/core-data/src/resolvers.js#L107-L111

I wonder what is the right level of abstraction here. Core-data could be selfish and introduce an action like `setEntityResolutionError`, but that name is already very telling. Resolution is a `@wordpress/data` concept, so perhaps there could be a meta action available like `setResolutionError`, similarly to how we have `finishResolution`? 

Taking it even further, we lack try/catch around the resolver call:

https://github.com/WordPress/gutenberg/blob/70730e61ba3148ebfa2be125d17c8138ab16392c/packages/data/src/redux-store/index.js#L413-L432

This sends all exceptions straight into the browser's console. I don't see how that's useful. Perhaps we could wrap it in try/catch and handle exceptions at that level? cc @gziolo @youknowriad

Reply:
I agree that error handling in core-data is not great at the moment, we need to add the errors to the resolution metadata state, add some selectors for it and yeah seems like a good addition to `useEntityRecord` or `useQuerySelect` as well.

Reply:
I started some explorations in https://github.com/WordPress/gutenberg/pull/38669 – let's move the discussion there